File: InMoto/Services/RouteGeometry.swift
import Foundation
import CoreLocation

/// Geometria "appiattita" di una NavigationRoute: tutte le tratte in un unico
/// polyline con distanze progressive. È la base del map matching: la posizione
/// GPS viene proiettata sulla linea e l'avanzamento (tappe, step, annunci)
/// deriva dalla progressione in metri lungo il percorso, mai da soglie puntuali
/// che in moto si saltano facilmente.
struct RouteGeometry {

    struct FlatStep {
        let legIndex: Int
        let progress: Double                    // metri dall'inizio del percorso
        let instruction: String
        let coordinate: CLLocationCoordinate2D
    }

    struct Match {
        let pointIndex: Int                     // primo punto del segmento agganciato
        let coordinate: CLLocationCoordinate2D  // proiezione sulla linea
        let progress: Double                    // metri dall'inizio
        let distanceFromRoute: Double           // distanza perpendicolare dalla linea
    }

    let points: [CLLocationCoordinate2D]
    let cumulative: [Double]        // metri dall'inizio, per ogni punto
    let waypointProgress: [Double]  // metri dall'inizio, per ogni waypoint
    let steps: [FlatStep]           // ordinati per progress

    var totalMeters: Double { cumulative.last ?? 0 }

    init(route: NavigationRoute) {
        var pts: [CLLocationCoordinate2D] = []
        var cum: [Double] = []
        var wpProg: [Double] = [0]
        var legRanges: [ClosedRange<Int>] = []

        for leg in route.legs {
            let firstIdx = max(0, pts.count - 1)
            for c in leg.polylineCoordinates {
                // Le tratte consecutive condividono il punto di giunzione: salta i duplicati
                if let last = pts.last,
                   abs(last.latitude - c.latitude) < 1e-7,
                   abs(last.longitude - c.longitude) < 1e-7 { continue }
                if let last = pts.last {
                    cum.append(cum[cum.count - 1] + Self.distance(last, c))
                } else {
                    cum.append(0)
                }
                pts.append(c)
            }
            legRanges.append(firstIdx...max(firstIdx, pts.count - 1))
            wpProg.append(cum.last ?? 0)
        }

        var flat: [FlatStep] = []
        for (li, leg) in route.legs.enumerated() {
            let range = legRanges[li]
            for step in leg.steps {
                let prog: Double
                if range.lowerBound < range.upperBound,
                   let m = Self.bestMatch(step.coordinate, points: pts, cumulative: cum,
                                          segments: range.lowerBound..<range.upperBound) {
                    prog = m.progress
                } else {
                    prog = wpProg[li]
                }
                flat.append(FlatStep(legIndex: li, progress: prog,
                                     instruction: step.instruction,
                                     coordinate: step.coordinate))
            }
        }
        flat.sort { $0.progress < $1.progress }

        self.points = pts
        self.cumulative = cum
        self.waypointProgress = wpProg
        self.steps = flat
    }

    /// Proietta una coordinata sul percorso. Cerca prima in una finestra attorno
    /// all'ultimo aggancio (evita di saltare su tratti vicini ma futuri/passati,
    /// es. anelli o nodi in comune tra tragitti); se fallisce estende la ricerca
    /// a tutto il percorso preferendo agganci coerenti con la progressione fatta.
    func match(_ coordinate: CLLocationCoordinate2D,
               nearPointIndex: Int?,
               minProgress: Double) -> Match? {
        guard points.count >= 2 else { return nil }
        let lastPoint = points.count - 1

        if let i = nearPointIndex {
            let lo = max(0, i - 40)
            let hi = min(lastPoint, i + 400)
            if lo < hi,
               let m = Self.bestMatch(coordinate, points: points, cumulative: cumulative,
                                      segments: lo..<hi),
               m.distanceFromRoute < 70 {
                return m
            }
        }

        // Ricerca "in avanti": solo sui tratti non ancora superati
        if minProgress > 0 {
            let startIdx = cumulative.firstIndex { $0 >= minProgress } ?? 0
            if startIdx < lastPoint,
               let m = Self.bestMatch(coordinate, points: points, cumulative: cumulative,
                                      segments: startIdx..<lastPoint),
               m.distanceFromRoute < 70 {
                return m
            }
        }

        return Self.bestMatch(coordinate, points: points, cumulative: cumulative,
                              segments: 0..<lastPoint)
    }

    /// Indice della tratta corrente data la progressione (un nodo conta come
    /// raggiunto già `reachThreshold` metri prima, per annunciare in tempo)
    func legIndex(forProgress p: Double, reachThreshold: Double = 30) -> Int {
        guard waypointProgress.count > 2 else { return 0 }
        var idx = 0
        for j in 1..<(waypointProgress.count - 1) where p + reachThreshold >= waypointProgress[j] {
            idx = j
        }
        return idx
    }

    /// Prossima manovra non ancora superata
    func upcomingStepIndex(afterProgress p: Double) -> Int? {
        steps.firstIndex { $0.progress > p + 8 }
    }

    // MARK: - Helpers geometrici

    static func distance(_ a: CLLocationCoordinate2D, _ b: CLLocationCoordinate2D) -> Double {
        CLLocation(latitude: a.latitude, longitude: a.longitude)
            .distance(from: CLLocation(latitude: b.latitude, longitude: b.longitude))
    }

    private static func bestMatch(_ c: CLLocationCoordinate2D,
                                  points: [CLLocationCoordinate2D],
                                  cumulative: [Double],
                                  segments: Range<Int>) -> Match? {
        guard !segments.isEmpty else { return nil }
        // Proiezione planare locale: sufficiente per distanze di pochi km
        let kLat = 111_132.0
        let kLon = 111_320.0 * cos(c.latitude * .pi / 180)

        var best: Match?
        for i in segments {
            let a = points[i], b = points[i + 1]
            let ax = (a.longitude - c.longitude) * kLon
            let ay = (a.latitude  - c.latitude)  * kLat
            let bx = (b.longitude - c.longitude) * kLon
            let by = (b.latitude  - c.latitude)  * kLat
            let dx = bx - ax, dy = by - ay
            let lenSq = dx * dx + dy * dy
            let t = lenSq > 0 ? min(1, max(0, -(ax * dx + ay * dy) / lenSq)) : 0
            let px = ax + dx * t, py = ay + dy * t
            let dist = (px * px + py * py).squareRoot()
            if best == nil || dist < best!.distanceFromRoute {
                best = Match(
                    pointIndex: i,
                    coordinate: CLLocationCoordinate2D(
                        latitude:  a.latitude  + (b.latitude  - a.latitude)  * t,
                        longitude: a.longitude + (b.longitude - a.longitude) * t),
                    progress: cumulative[i] + (cumulative[i + 1] - cumulative[i]) * t,
                    distanceFromRoute: dist)
            }
        }
        return best
    }
}

File: android/app/src/main/java/com/divito/inmoto/data/RouteGeometry.kt
package com.divito.inmoto.data

import com.divito.inmoto.model.GeoPoint
import com.divito.inmoto.model.NavigationRoute
import kotlin.math.PI
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Geometria "appiattita" di una [NavigationRoute]: tutte le tratte in un unico
 * polyline con distanze progressive. È la base del map matching: la posizione GPS
 * viene proiettata sulla linea e l'avanzamento (tappe, step, annunci) deriva dalla
 * progressione in metri lungo il percorso, mai da soglie puntuali che in moto si
 * saltano facilmente. Port 1:1 di RouteGeometry.swift.
 */
class RouteGeometry(route: NavigationRoute) {

    data class FlatStep(
        val legIndex: Int,
        val progress: Double,          // metri dall'inizio del percorso
        val instruction: String,
        val coordinate: GeoPoint,
    )

    data class Match(
        val pointIndex: Int,           // primo punto del segmento agganciato
        val coordinate: GeoPoint,      // proiezione sulla linea
        val progress: Double,          // metri dall'inizio
        val distanceFromRoute: Double, // distanza perpendicolare dalla linea
    )

    val points: List<GeoPoint>
    val cumulative: List<Double>       // metri dall'inizio, per ogni punto
    val waypointProgress: List<Double> // metri dall'inizio, per ogni waypoint
    val steps: List<FlatStep>          // ordinati per progress

    val totalMeters: Double get() = cumulative.lastOrNull() ?: 0.0

    init {
        val pts = ArrayList<GeoPoint>()
        val cum = ArrayList<Double>()
        val wpProg = ArrayList<Double>().apply { add(0.0) }
        val legRanges = ArrayList<IntRange>() // primo..ultimo indice punto (inclusivi)

        for (leg in route.legs) {
            val firstIdx = max(0, pts.size - 1)
            for (c in leg.polylineCoordinates) {
                // Le tratte consecutive condividono il punto di giunzione: salta i duplicati
                val last = pts.lastOrNull()
                if (last != null && abs(last.lat - c.lat) < 1e-7 && abs(last.lon - c.lon) < 1e-7) continue
                if (last != null) cum.add(cum[cum.size - 1] + distance(last, c)) else cum.add(0.0)
                pts.add(c)
            }
            legRanges.add(firstIdx..max(firstIdx, pts.size - 1))
            wpProg.add(cum.lastOrNull() ?: 0.0)
        }

        val flat = ArrayList<FlatStep>()
        route.legs.forEachIndexed { li, leg ->
            val range = legRanges[li]
            for (step in leg.steps) {
                val prog = if (range.first < range.last) {
                    bestMatch(step.point, pts, cum, range.first until range.last)?.progress ?: wpProg[li]
                } else {
                    wpProg[li]
                }
                flat.add(FlatStep(li, prog, step.instruction, step.point))
            }
        }
        flat.sortBy { it.progress }

        points = pts
        cumulative = cum
        waypointProgress = wpProg
        steps = flat
    }

    /**
     * Proietta una coordinata sul percorso. Cerca prima in una finestra attorno
     * all'ultimo aggancio; se fallisce estende la ricerca in avanti e poi a tutto
     * il percorso. Port di RouteGeometry.match.
     */
    fun match(coordinate: GeoPoint, nearPointIndex: Int?, minProgress: Double): Match? {
        if (points.size < 2) return null
        val lastPoint = points.size - 1

        if (nearPointIndex != null) {
            val lo = max(0, nearPointIndex - 40)
            val hi = min(lastPoint, nearPointIndex + 400)
            if (lo < hi) {
                val m = bestMatch(coordinate, points, cumulative, lo until hi)
                if (m != null && m.distanceFromRoute < 70) return m
            }
        }

        // Ricerca "in avanti": solo sui tratti non ancora superati
        if (minProgress > 0) {
            val startIdx = cumulative.indexOfFirst { it >= minProgress }.let { if (it < 0) 0 else it }
            if (startIdx < lastPoint) {
                val m = bestMatch(coordinate, points, cumulative, startIdx until lastPoint)
                if (m != null && m.distanceFromRoute < 70) return m
            }
        }

        return bestMatch(coordinate, points, cumulative, 0 until lastPoint)
    }

    /** Indice della tratta corrente data la progressione (nodo "raggiunto" già
     * `reachThreshold` metri prima, per annunciare in tempo). */
    fun legIndex(progress: Double, reachThreshold: Double = 30.0): Int {
        if (waypointProgress.size <= 2) return 0
        var idx = 0
        for (j in 1 until waypointProgress.size - 1) {
            if (progress + reachThreshold >= waypointProgress[j]) idx = j
        }
        return idx
    }

    /** Prossima manovra non ancora superata. */
    fun upcomingStepIndex(afterProgress: Double): Int? {
        val i = steps.indexOfFirst { it.progress > afterProgress + 8 }
        return if (i < 0) null else i
    }

    /** Direzione di marcia (gradi, 0=N) lungo il percorso alla progressione data,
     * per la camera course-up. Guarda un po' avanti per essere stabile. */
    fun bearingAtProgress(progress: Double, lookahead: Double = 35.0): Double? {
        if (points.size < 2) return null
        val iA = cumulative.indexOfFirst { it >= progress }.let { if (it < 0) points.size - 1 else it }
        val iB = cumulative.indexOfFirst { it >= progress + lookahead }.let { if (it < 0) points.size - 1 else it }
        val a = points[min(iA, points.size - 1)]
        val b = points[min(max(iB, iA + 1), points.size - 1)]
        if (abs(a.lat - b.lat) < 1e-9 && abs(a.lon - b.lon) < 1e-9) return null
        return bearing(a, b)
    }

    companion object {
        /** Distanza in metri (Haversine). */
        fun distance(a: GeoPoint, b: GeoPoint): Double {
            val r = 6_371_000.0
            val dLat = Math.toRadians(b.lat - a.lat)
            val dLon = Math.toRadians(b.lon - a.lon)
            val la1 = Math.toRadians(a.lat)
            val la2 = Math.toRadians(b.lat)
            val h = sin(dLat / 2).pow(2) + cos(la1) * cos(la2) * sin(dLon / 2).pow(2)
            return 2 * r * asin(min(1.0, sqrt(h)))
        }

        /** Rotta iniziale in gradi (0=N) da a verso b. */
        fun bearing(a: GeoPoint, b: GeoPoint): Double {
            val la1 = Math.toRadians(a.lat)
            val la2 = Math.toRadians(b.lat)
            val dLon = Math.toRadians(b.lon - a.lon)
            val y = sin(dLon) * cos(la2)
            val x = cos(la1) * sin(la2) - sin(la1) * cos(la2) * cos(dLon)
            return (Math.toDegrees(atan2(y, x)) + 360) % 360
        }

        private fun bestMatch(
            c: GeoPoint,
            points: List<GeoPoint>,
            cumulative: List<Double>,
            segments: IntRange,
        ): Match? {
            if (segments.isEmpty()) return null
            // Proiezione planare locale: sufficiente per distanze di pochi km
            val kLat = 111_132.0
            val kLon = 111_320.0 * cos(c.lat * PI / 180)

            var best: Match? = null
            for (i in segments) {
                val a = points[i]
                val b = points[i + 1]
                val ax = (a.lon - c.lon) * kLon
                val ay = (a.lat - c.lat) * kLat
                val bx = (b.lon - c.lon) * kLon
                val by = (b.lat - c.lat) * kLat
                val dx = bx - ax
                val dy = by - ay
                val lenSq = dx * dx + dy * dy
                val t = if (lenSq > 0) min(1.0, max(0.0, -(ax * dx + ay * dy) / lenSq)) else 0.0
                val px = ax + dx * t
                val py = ay + dy * t
                val dist = sqrt(px * px + py * py)
                val cur = best
                if (cur == null || dist < cur.distanceFromRoute) {
                    best = Match(
                        pointIndex = i,
                        coordinate = GeoPoint(
                            lat = a.lat + (b.lat - a.lat) * t,
                            lon = a.lon + (b.lon - a.lon) * t,
                        ),
                        progress = cumulative[i] + (cumulative[i + 1] - cumulative[i]) * t,
                        distanceFromRoute = dist,
                    )
                }
            }
            return best
        }
    }
}
